ipykernel/kernelbase.py

```diff
@@ -553,9 +553,8 @@ async def start(self, *, task_status: TaskStatus = TASK_STATUS_IGNORED) -> None:
                     tg.start_soon(self.shell_main, None)
 
     def stop(self):
-        if not self._eventloop_set.is_set():
-            # Stop the async task that is waiting for the eventloop to be set.
-            self._eventloop_set.set()
+        # Stop the async task that is waiting for the eventloop to be set.
+        self._eventloop_set.set()
```

Comment:
you possibly want to use https://anyio.readthedocs.io/en/stable/threads.html#calling-asynchronous-code-from-an-external-thread

Reply:
or if that's overkill you could use:

```python
def current_token() -> tuple[Literal["asyncio"], AbstractEventLoop] | tuple[Literal["trio"], trio.lowlevel.TrioToken]:
    lib = sniffio.current_async_library()
    if lib == "asyncio":
        return lib, asyncio.get_running_loop()
    if lib == "trio:
        return lib, trio.lowlevel.current_trio_token()

def from_thread_run_sync[**P](token: tuple[Literal["asyncio"], AbstractEventLoop] | tuple[Literal["trio"], trio.lowlevel.TrioToken], fn: Callable[P, object], /, *args: P.args, **kwargs: P.kwargs) -> None:
    lib, tok = token
    if lib == "asyncio":
        tok.call_soon_threadsafe(functools.partial(fn, *args, **kwargs)
        return
    if lib == "trio":
        trio.from_thread.run_sync(functools.partial(fn, *args, **kwargs), trio_token=tok)
```


```python
    async def _wait_to_enter_eventloop(self):
        self.kernel._token = current_token()
        await self.kernel._eventloop_set.wait()
        await self.kernel.enter_eventloop()
```

```python
def stop(self):
    from_thread_run_sync(self._token, self._eventloop_set.set)
```